### PR TITLE
Feature/verification key update txn version

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
Explain your changes:
This PR implements this RFC: https://github.com/MinaProtocol/mina/blob/9577ad689a8e4d4f97e1d0fc3d26e20219f4abd1/rfcs/0051-verification-key-permissions.md
The implementation changes the `Permissions.t`'s `set_verification_key` field to make it be `'controller * 'txn_version`.
It adds a well formedness check against the account updates that only accept version that's has the same transaction version as the current one;
it also alters the zkapp_command_logic so that when user tries to modify verification key if the account permission is set to `Proof` or `Impossible` with an older version, then in this case the permission for verification key update would fallback to signature.

Explain how you tested your changes:
Corresponding tests are added for transaction snark and zkapp_command and 2 of the integration tests are modified:
1. hard-fork integration test; I added 2 zkapp accounts to the genesis ledger with `Proof` and `Impossible` as their permissions for verification key update with older version. This is to test the "fallback" case on `Signature` authorization
2. verification-key-update test; here I set up 2 zkapp accounts and changed their permissions to `Proof` and `Impossible` and checked they are still well-behaved.

The merge conflict is caused by:
1. `berkeley_migration` app is already removed in `develop` and this PR has some modification on it to make things compile
2. `snarkyjs` submodule is removed in `develop` and this PR has some additional commits to this submodule

Checklist:

- [ ] Dependency versions are unchanged
  - Notify Velocity team if dependencies must change in CI
- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them

* Closes #0000

### CODE REVIEW

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

File: src/app/archive/lib/processor.ml [hunk 1]
```diff
@@ -622,15 +664,33 @@ module Zkapp_permissions = struct
...
+      Protocol_versions.find_txn_version (module Conn) ~transaction:txn_version
+    in
+    ( match versions with
+    | Ok [] ->
+        [%log error]
+          ~metadata:[ ("perms", Permissions.to_yojson perms) ]
+          "No transaction version exists for the permission"
+    | Ok _ ->
+        ()
+    | Error _ ->
```
Comment (psteckler) [inline src/app/archive/lib/processor.ml:682]: should the error also be printed?

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by Trivo25** [comment]
!ci-build-me

**Review by Trivo25 — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

File: src/lib/mina_base/zkapp_command.ml [hunk 2]
```diff
@@ -1955,6 +1955,14 @@ let has_zero_vesting_period t =
       | Set { vesting_period; _ } ->
           Mina_numbers.Global_slot_span.(equal zero) vesting_period )
 
+let is_incompatible_version t =
+  Call_forest.exists t.account_updates ~f:(fun p ->
+      match p.body.update.permissions with
+      | Keep ->
+          false
+      | Set { set_verification_key = auth, txn_version; _ } ->
```
Comment (nholland94) [inline src/lib/mina_base/zkapp_command.ml:1963]: Do you not get a warning for this unused `auth` value? If so, we should work to re-enable the unused value warnings for this file (in a separate PR).

**Review by nholland94 — CHANGES_REQUESTED** [review]
Overall LGTM, but I would really like to see more clarity in how all of this is surface to the transaction logic. That could mean just a better name for this `fallback` function, or moving this fallback logic explicitly inside of the transaction logic. But just this generic `fallback` function makes it a bit hard to understand what's going on within the transaction logic code, which is something I think we should optimize for in our changes there.

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-nightly-me

**Review by nholland94 — APPROVED** [review]

**Review by mrmr1993 — CHANGES_REQUESTED** [review]
LGTM, but please update the proof caches for the transaction snark tests now that the circuits have changed. You can find some details on how to do that [here](https://github.com/MinaProtocol/mina/tree/feature/verification-key-update-txn-version/src/lib/transaction_snark/test#readme).

File: src/app/test_executive/block_production_priority.ml [hunk 3]
```diff
@@ -24,20 +24,16 @@ module Make (Inputs : Intf.Test.Inputs_intf) = struct
     { default with
       requires_graphql = true
     ; genesis_ledger =
-        [ { Test_Account.account_name = "receiver-key"
-          ; balance = "9999999"
-          ; timing = Untimed
-          }
-        ; { account_name = "empty-bp-key"; balance = "0"; timing = Untimed }
-        ; { account_name = "snark-node-key"; balance = "0"; timing = Untimed }
+        (let open Test_account in
```
[the remark this answers is not in the document]
Comment (ghost-not-in-the-shell) [reply src/app/test_executive/block_production_priority.ml:27]: I modified the `Test_account` to do testing of verification key update in integration tests. Unfortunately this change propagates to other integration tests.

 I do realize that the changes to protocol and the changes to the tests seem like rather unrelated in this PR. Maybe I should also have a chain of PRs that separate the change itself and the change to the integration tests to make review easier.

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Review by mrmr1993 — APPROVED** [review]

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by ghost-not-in-the-shell** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by deepthiskumar** [comment]
!ci-build-me

**Comment by deepthiskumar** [comment]
!ci-nightly-me

**Comment by deepthiskumar** [comment]
https://buildkite.com/o-1-labs-2/mina-end-to-end-nightlies/builds/1145

**Comment by deepthiskumar** [comment]
Waiting for all the intg tests to pass

**Comment by deepthiskumar** [comment]
!approved-for-mainnet

**Pull request merged**